accepted/single-file/design.md

```diff
@@ -66,225 +32,254 @@ Publishing to a single file can be triggered by adding the following property to
+| SuperHost  | StaticHost with CoreCLR runtime components statically linked | Self-contained single-file apps on Unix systems              |
+
+Ideally, we should use the SuperHost for self-contained single-file apps on Windows too. However, due to certain limitations in debugging experience and the ability to collect Watson dumps, etc., the CoreCLR libraries are left on disk beside the app.
+
+The files typically published for self-contained apps on Windows are:
+
+* The `App.Exe`-- StaticHost along with assemblies and configuration files bundled as a single-file app.
+* `coreclr.dll`(runtime), `clrjit.dll` (JIT compiler), and `clrcompression.dll` (native counterpart of BCL for certain compression algorithms).
+    * It may be possible to link these DLLs together as CoreCLR.dll, but this work is not prioritized, in deference to building the full super-host once debugging framework supports it.
+* mscordaccore.dll (to enable Watson dumps)
+
+Certain additional binaries may be optionally included with the app to enable debugging scenarios. 
+The work to not require these binaries as separate files on disk alongside the app is ongoing.
+
+* For F5 debugging on VS and VS-Core: `mscordbi.dll`, `libmscordbi.so`
+* For Linux mini-dumps: `createdump`, `libmscordaccore.so` 
+* For ETW / LTTng: `clretwrc.dll`, `libcoreclrtraceptprovider.so` 
+* For exception stack trace source information, error string resources on Windows: `Microsoft.DiaSymReader.Native.amd64.dll`, `mscorrc.debug.dll`, `mscorrc.dll`
+
+When targeting `win7` platform, several additional DLLs are necessary (`api-*.dll`) to handle API compatibility. These files must be alongside the AppHost for the app to even start execution.
+Therefore, we propose that targeting `win7-*` should not be supported when publishing apps as a single-file.
+
+## The Runtime
+
+### Dependency Resolution
+
+In order to resolve assemblies, the embedded resources are probed first, followed by other probing paths encoded in `TRUSTED_PLATFORM_ASSEMBLIES`. 
+
+### PEImage loader
+
+* IL assemblies are loaded directly from the bundle.
+    * The portion of the single-file bundle containing the required assembly is memory mapped, and the contents are appropriately interpreted by the runtime.
+
+* ReadyToRun Assemblies
+    * On Linux, ReadyToRun assemblies are loaded directly from bundle. The various sections in the PE file are mapped at appropriate addresses, and offsets are fixed up.
+    * On MAC, ReadyToRun assemblies are loaded similar to Linux. However, Mojave hardened runtime doesn't allow executable mappings of a file. Therefore, the contents of an assembly are read from the bundle into pre-allocated executable memory.
+    * On Windows, due to certain limitations in memory mapping routines described below, ReadyToRun assemblies are loaded by memory mapping the file and copying sections to appropriate offsets.
```

Comment:
Note that we can do better here by laying out the sections in the file such that this copying is not necessary.

Reply:
Ok, I see you are covering this below.